service/node.go

```diff
@@ -504,9 +519,126 @@ func (s *service) NodeGetInfo(
 	}, nil
 }
 
-func (s *service) NodeGetVolumeStats(
-	ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+// NodeGetVolumeStats will check the status of a volume given its ID and path
+// if volume is healthy, stats on volume usage will be returned
+// if volume is unhealthy, a message will be returned detailing the issue
+// To determine if volume is healthy, this method checks: volume known to array, volume known to SDC, volume path readable, and volume path mounted
+// Note: kubelet only calls this method when feature gate: CSIVolumeHealth=true
+func (s *service) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+
+	csiVolID := req.GetVolumeId()
+	volPath := req.GetVolumePath()
+	mounted := false
+	healthy := true
+	message := ""
+
+	//validate params first, make sure neither field is empty
+	if len(volPath) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "no volume Path provided")
+	}
+
+	if len(csiVolID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "no volume ID  provided")
+	}
+
+	//check if volume exists
+
+	volID := getVolumeIDFromCsiVolumeID(csiVolID)
+	systemID := s.getSystemIDFromCsiVolumeID(csiVolID)
+
+	if systemID == "" {
+		// use default system
+		systemID = s.opts.defaultSystemID
+	}
+
+	_, err := s.getSDCMappedVol(volID, systemID, 30)
+	if err != nil {
+		//volume not known to SDC, next check if it exists at all
+		_, _, err := s.listVolumes(systemID, 0, 0, false, false, volID, "")
+		if err != nil && strings.Contains(err.Error(), sioGatewayVolumeNotFound) {
+			message = fmt.Sprintf("Could not get volume with ID '%s' from array %s due to error: %s", volID, systemID, err)
+
+		} else if err != nil {
+			//err wasn't not found, return error
+			return nil, err
+		}
+		//volume was found, but was not known to SDC. This is abnormal.
+		healthy = false
+		if message == "" {
+			message = fmt.Sprintf("volume: %s was not mapped to host: %v", volID, err)
+		}
+
+	}
+
+	//check if volume path is accessible
+	if healthy {
+		_, err = os.ReadDir(volPath)
+		if err != nil && healthy {
+			healthy = false
+			message = fmt.Sprintf("volume path: %s is not accessible: %v", volPath, err)
+		}
+	}
+
+	if healthy {
+
+		//check if path is mounted on node
+		mounts, err := getPathMounts(volPath)
+		if len(mounts) > 0 {
+			for _, m := range mounts {
+				if m.Path == volPath {
+					Log.Infof("volPath: %s is mounted", volPath)
+					mounted = true
+				}
+			}
+		}
+		if len(mounts) == 0 || mounted == false || err != nil {
+			healthy = false
+			message = fmt.Sprintf("volPath: %s is not mounted: %v", volPath, err)
```

Comment:
Is there a way to make the message more explicit? If this fails because of an empty mounts list or because the last volume from the 588-591 loop was not mounted, then we just get no error message after the colon, which seems a little cryptic. 

Additionally, it seems like there might be a hole in how we do this -- what if there is a list of three volumes, and the first and third are mounted, but the second one is not? Then we won't record a message here even though we probably should.

Reply:
We only check 1 volume when the method is called, so we don't need to worry about a mix of mounted/unmounted volumes. 

An empty mount list means the volume is not mounted, which is why the message is returned as "not mounted"

Reply:
If that's the case, why have the for loop at all? It makes it look like there is a series of mounts

Reply:
there is only 1 volume, but 1 volume can potentially have more than 1paths  mounted on it. So we need to check to make sure. 
Even though in most cases, 1 volume will only have 1 mount